### PR TITLE
Add Serialization Module

### PR DESCRIPTION
**Pull request by PradyumnaKrishna**
### Description of the changes being introduced by the pull request:
Metadata classes and other similar classes requires a method to
serialize and deserialize their objects into json, bytes, and
other forms. Hence, to provide a common interface to serialize
and deserialize is a better approach.

BaseSerializer and BaseDeserializer are introduced having
method serialize and deserialize respectively. These methods will
serialize class instance to raw bytes and vice versa.

SerializationError and DeserializationError will be raised on error
occurrences.

### Please verify and check that the pull request fulfils the following requirements:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

### CODE REVIEW

**Comment by lukpueh** [comment]
This is extremely cool! I like how you provide common serialization-related methods in the `Serializable` class.

Do you think it's worth to define (in code) the expectations with have for classes that use `JSONSerializer` (must implement `to_dict`) and `JSONDeserializer` (must implement `from_dict`)?
 
It would be cool, if we could use `typing.Protocol`. It seems like the "right" way to define those expectations, without requiring classes like `Metadata` or `Envelope` to subclass e.g. a `JSONSerializable`, or `JSONDeserializable` abstract class.

File: securesystemslib/metadata.py [hunk 1]
```diff
@@ -161,3 +171,28 @@ def verify(self, keys: List[Key], threshold: int) -> Dict[str, Key]:
...
+
+    def deserialize_payload(
+        self,
+        class_type: Any,
+        deserializer: Optional[BaseDeserializer] = None,
+    ) -> Any:
+        """Parse DSSE payload.
+
+        Arguments:
+            class_type: A class having a from_dict method.
```
Comment (lukpueh) [inline securesystemslib/metadata.py]: This only has to be `from_dict` if we use default `JSONDeserializer`. For other deserializers `class_type` may not even be needed. Should we make it optional?

**Review by lukpueh — COMMENTED** [review]
Implementation-wise this looks impeccable! I really only have a couple of high-level thoughts to share. IMO the following classes of this PR would be better candidates for `typing.Protocol` than `abc`:
- BaseSerializer
- BaseDeserializer
- JSONSerializable

Mind adding a TODO comment or creating a ticket for post-Python 3.7 times?

I also hope that this does not turn out too "engineered". It really looks smart and we definitely want to make the code more robust and enforce expectations, such as "json deserialize requires the used class to implement a `from_dict` method", but we also don't want the code to look like java.

I'd appreciate an assessment from one of the  theupdateframework/python-tuf#1279-reviewers - cc @joshuagl, @jku. FYI: This is @PradyumnaKrishna's take on python-tuf's serialization subpackage, making it more re-usable/flexible in terms of which classes it can be used with, e.g. an `Envelope` class, which is the DSSE equivalent of python-tuf's Metadata class. The goal is to use this implementation in python-tuf too at some point.

File: securesystemslib/metadata.py (hunk 1 above)
Comment (PradyumnaKrishna) [reply securesystemslib/metadata.py]: `BaseDeserializer` requires a class type to store the bytes into a containerized form. That's why the class type will be required by every deserializer.

Comment (lukpueh) [reply securesystemslib/metadata.py]: My comment about `from_dict` ist still true. Not every `deserializer` might require the class to have a `from_dict` method. What about something like...
```suggestion
            class_type: The class to  be deserialized. If the default deserializer is
                used, it must implement ``JSONSerializable``.
```

**Review by lukpueh — APPROVED** [review]
💯

**Pull request merged**